metricbeat/module/elasticsearch/cluster_stats/data_xpack.go

```diff
@@ -0,0 +1,238 @@
+					"min": c.Int("min"),
+					"max": c.Int("max"),
+					"avg": c.Int("avg"),
+				}),
+			}),
+			"jvm": c.Dict("jvm", s.Schema{
+				"max_uptime_in_millis": c.Int("max_uptime_in_millis"),
+				"mem": c.Dict("mem", s.Schema{
+					"heap_used_in_bytes": c.Int("heap_used_in_bytes"),
+					"heap_max_in_bytes":  c.Int("heap_max_in_bytes"),
+				}),
+				"threads": c.Int("threads"),
+			}),
+			"fs": c.Dict("fs", s.Schema{
+				"total_in_bytes":     c.Int("total_in_bytes"),
+				"free_in_bytes":      c.Int("free_in_bytes"),
+				"available_in_bytes": c.Int("available_in_bytes"),
+			}),
+		}),
+	}
+)
+
+func passthruField(fieldPath string, sourceData, targetData common.MapStr) error {
+	fieldValue, err := sourceData.GetValue(fieldPath)
+	if err != nil {
+		return elastic.MakeErrorForMissingField(fieldPath, elastic.Elasticsearch)
+	}
+
+	targetData.Put(fieldPath, fieldValue)
+	return nil
+}
+
+func eventMappingXPack(r mb.ReporterV2, m *MetricSet, content []byte) error {
+	var data map[string]interface{}
+	err := json.Unmarshal(content, &data)
+	if err != nil {
+		return err
+	}
+
+	clusterStats, err := clusterStatsSchema.Apply(data)
+	if err != nil {
+		return err
+	}
+
+	dataMS := common.MapStr(data)
+
+	passthruFields := []string{
+		"indices.segments.file_sizes", // object with dynamic keys
+		"nodes.versions", // array of strings
+		"nodes.os.names", // array of objects
+		"nodes.jvm.versions", // array of objects
+		"nodes.plugins", // array of objects
+		"nodes.network_types", // object with dynamic keys
+	}
+	for _, fieldPath := range passthruFields {
+		if err = passthruField(fieldPath, dataMS, clusterStats); err != nil {
+			return err
+		}
+	}
+
+	clusterName, ok := data["cluster_name"].(string)
+	if !ok {
+		return elastic.MakeErrorForMissingField("cluster_name", elastic.Elasticsearch)
+	}
+
+	info, err := elasticsearch.GetInfo(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	license, err := elasticsearch.GetLicense(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	// TODO: Inject `cluster_needs_tls` field under license object
+
+	clusterState, err := elasticsearch.GetClusterState(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	if err = passthruField("status", dataMS, clusterState); err != nil {
+		return err
+	}
+
+	// TODO: Compute and inject `node_hash` field under clusterState object
+
+	stackStats, err := elasticsearch.GetStackStats(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	// TODO: Inject `apm.found` field under stackStats object
+
+	event := mb.Event{}
+	event.RootFields = common.MapStr{
+		"cluster_uuid":  info.ClusterID,
+		"cluster_name":  clusterName,
+		"timestamp":     common.Time(time.Now()),
+		"interval_ms":   m.Module().Config().Period / time.Millisecond,
+		"type":          "cluster_stats",
+		"license":       license,
+		"version":       info.Version.Number,
+		"cluster_stats": clusterStats,
```

Comment:
@ruflin As you can see here, the document that is indexed into monitoring contains not only `clusterStats` but also `clusterState` (an abridged version of it) and `stackStats`.

For [`clusterState`](https://github.com/elastic/beats/pull/7810/files#diff-5c3e5a2da8527b611dff0cc47e744173R202) and [`stackStats`](https://github.com/elastic/beats/pull/7810/files#diff-5c3e5a2da8527b611dff0cc47e744173R213) we are simply calling the corresponding Elasticsearch APIs, doing a tiny bit of massaging of the response data, and then passing the resulting structure through over here. We are definitely not doing a complete schema conversion like we are doing for [`clusterStats`](https://github.com/elastic/beats/pull/7810/files#diff-5c3e5a2da8527b611dff0cc47e744173R34).

It makes me wonder: should we treat `clusterStats` special over here? Or conversely, should we do complete schema conversions for `clusterState` and `stackStats` too?

Reply:
How is the internal reporting in ES happening? Is it just taking all the existing fields? If yes then we should do the same here.

Reply:
Yes, it just takes all the existing fields and passes them through. Here is the relevant code:

https://github.com/elastic/elasticsearch/blob/b87f3062b77cab7888e9037b4996f2c26db12816/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDoc.java#L141-L147